system/lib/pthread/pthread_create.c

```diff
@@ -236,7 +236,7 @@ void _emscripten_thread_exit(void* result) {
   // object and we are done.
   if (state == DT_DETACHED) {
     self->detach_state = DT_EXITED;
-    __pthread_detached_exit();
+    __emscripten_thread_cleanup(self);
```

Comment:
I was thinking about doing this .. but I was slightly worried about free'ing the stack being used by the current thread.   

Seems a little dangerous.  I don't come up with better solution yet but I was thinking maybe we could instead call this from worker.js?  (i.e.  once the stack is unwound all the way)?

Reply:
IIUC, `__emscripten_thread_free_data` was also called by the `detachedExit`/`__pthread_detached_exit`, so this PR should be completely non-functional (except that it now sends the `cleanupThread` command, which will also eventually call `returnWorkerToPool`/`__emscripten_thread_free_data`).

Reply:
IIUC, In the current code, when `__pthread_detached_exit` is called the `cleanupThread` command is sent back to the main thread which ends up freeing the data, so the data is not freed on the exit'ing thread at all.

Reply:
Also, with this new PR, how is the detached thread ever returned to the worker pool?

Reply:
Oh wait, I see that `__emscripten_thread_cleanup` also posts a message back to the main thread.. so I agree this should be and NFC.  Thanks for clarifying!